File: Python/skyscrapers.py
# Write a Pyhton funtion that given a list of positive integers values
# returns how many times the maximum change
from random import shuffle

def count_tops(values: list[int]) -> int:
    max = values[0]
    max_changes = 1
    for val in values:
        if val > max:
            max_changes += 1
            max = val
    
    return max_changes

def main():
    values = list(range(1,6))
    shuffle(values)
    print(values)
    print(count_tops(values))

if __name__ == "__main__":
    main()
    

File: Haskell/skyscrapers.hs
-- Given a list of unique positive integers values...
    -- How many times does the maximum change?
    -- Adapt to Haskell, with tail recursion

skyscrapers :: (Num a, Ord a) => [a] -> a
skyscrapers [] = error "empty list"
skyscrapers x = skyscraperRecursion x []

skyscraperRecursion :: (Num a, Ord a) => [a] -> [a] -> a
skyscraperRecursion [] acc = fromIntegral (length acc)
skyscraperRecursion (x:xs) [] = skyscraperRecursion xs [x]
skyscraperRecursion (x:xs) acc 
 | x > (head acc) = skyscraperRecursion xs (x:acc)
 | otherwise = skyscraperRecursion xs acc

